torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py

```diff
@@ -374,6 +374,361 @@ def collect_bw_donated_buffer_idxs(
+                )
+
+                # Step 2) and 3) - Run joint graph passes and partitioner
+                new_fw_hop_gm, new_bw_hop_gm = aot_config.partition_fn(
+                    joint_hop_gm, [], num_fwd_outputs=num_fw_outputs
+                )
+
+                # Save the new forward and backward graph modules
+                new_hop_graphs[identifier].new_fw_hop_gm = new_fw_hop_gm
+                new_hop_graphs[identifier].new_bw_hop_gm = new_bw_hop_gm
+
+                # Save the number of symints and saved tensors
+                new_fw_out_nodes = new_fw_hop_gm.graph.find_nodes(op="output")[0].args[
+                    0
+                ]
+                extra_outputs = new_fw_out_nodes[num_fw_outputs:]
+                symint_outputs = [n for n in extra_outputs if is_sym_node(n)]
+
+                new_hop_graphs[identifier].new_num_sym_nodes = len(symint_outputs)
+                new_hop_graphs[identifier].new_num_saved_nodes = len(
+                    extra_outputs
+                ) - len(symint_outputs)
+
+                new_hop_graphs[identifier].partitioning_done = True
+
+    if not new_hop_graphs:
+        return joint_gm
+
+    # Step 3) Restitch the new fw and bw graphs back into the main graph.
+    #
+    # This is a very mechanical process. There are a quite of pieces that we
+    # need to connect together to make it work. Lets try to understand the
+    # problem statement first.
+    #
+    # For the forward graph, the signature of the old_fw_hop_gm is
+    #   inputs - (*primals)
+    #   outputs - (*fw_outs)
+    # Now the signature of the new_fw_hop_gm is
+    #   inputs - (*primals)     -- This is same
+    #   outputs - (*fw_outs, *saved_tensors)    - This is different
+    # At a high level, this is an easy transformation, in the new graph we just
+    # have to replace the old_fw_hop_gm with the new_fw_hop_gm. Everything else
+    # falls into place, because the input signature (i.e. args) is same. And
+    # even though output signature is different, fw_outs are still at the same
+    # indexes as before. So the forward of the `joint_gm` works nicely.
+    #
+    # Now, lets look at the backward hop graph. Old signature
+    #   inputs - (*primals, *tangents)
+    #   outputs - (*grad_outs, *fw_outs)
+    # New signature
+    #   inputs - (*saved_tensors, *tangents) -- Different
+    #   outputs - (*grad_outs)  -- Different
+    # Here both input and output signature change. The output signature handling
+    # is quite easy because the grads_out are sitting at the right place, so we
+    # dont have to do anything.
+    #
+    # For the input signature, we have to collect the saved tensors from the
+    # corresponding forward graph output. We collect all saved_tensors when we
+    # see the forward graph, and save it into a map and then later use it during
+    # the backward.
+
+    env = {}  # type: ignore[var-annotated]
+    new_graph = torch.fx.Graph()
```

Comment:
If I'm understanding this correctly, we are producing a new joint graph from the original joint graph. Why not modify the original joint graph in-place?

Reply:
Just did not think it that way. I can make it in-place.

Just curious, is there any benefit of in-place modifying vs creating a new graph?

Reply:
1) performance - creating a whole new joint graph is slower than just editing the invoke_subgraph's forward subgraph and backward subgraph
2) when creating a new graph, we need to define how to propagate some of the node.meta over. This is error-prone sometimes (I'm currently adding a field that does not always get propagated). Mutating the original graph in-place makes this less of a concern